### PR TITLE
[sfputilbase]: enhance the sfputilbase for transceiver monitoring

### PR DESCRIPTION
**Pull request by keboliu**
[sfputilbase]: enhance the sfputilbase for transceiver monitoring
    
    [List of changes]
    
    * add ability to only read and parse specific bytes from eeprom
    * add two API to read the transceiver info and dom info which intrested by SNMP
    * add a new API definiton: get_transceiver_change_event
    * fix a bug: wrong 'VendorOUI' offset for sff8472
    
        modified:   sff8436.py
        modified:   sff8472.py
        modified:   sfputilbase.py
    
    Signed-off-by Liu Kebo kebol@mellanox.com

### CODE REVIEW

File: sonic_sfp/sff8436.py [hunk 1]
```diff
@@ -364,6 +364,41 @@ class sff8436InterfaceId(sffbase):
                  'type'  : 'bitmap',
                  'decode': {}}}
 
+    sfp_type = {
+        'type':
+            {'offset':0,
```
Comment (jleveque) [inline sonic_sfp/sff8436.py]: Add space after `:` for consistent PEP8 style (`{'offset': 0,`).

There are many more occurrences. Please do this throughout the entire file.

File: sonic_sfp/sff8436.py [hunk 2]
```diff
@@ -364,6 +364,41 @@ class sff8436InterfaceId(sffbase):
...
+             'size':1,
+             'type' : 'enum',
```
Comment (jleveque) [inline sonic_sfp/sff8436.py]: Delete space before `:` for consistent PEP8 style (`{''type': 'enum',`).

There are many more occurrences. Please do this throughout the entire file.

File: sonic_sfp/sff8436.py [hunk 3]
```diff
@@ -923,6 +972,66 @@ def calc_rx_power(self, eeprom_data, offset, size):
...
+             'type': 'func',
+             'decode': {'func':calc_temperature}}
+        }
+
+    dom_module_voltage = {
+        'Vcc':
+            {'offset':0,
+             'size':2,
+             'type': 'func',
+             'decode': { 'func':calc_voltage}}
```
Comment (jleveque) [inline sonic_sfp/sff8436.py]: Delete space after `{` for consistent PEP8 style (`'decode': {'func':calc_voltage}}`).

There are many more occurrences. Please do this throughout the entire file.

File: sonic_sfp/sfputilbase.py [hunk 4]
```diff
@@ -268,15 +310,14 @@ def read_porttab_mappings(self, porttabfile):
 
         # Read the porttab file and generate dicts
         # with mapping for future reference.
-        #
-        # TODO: Refactor this to use the portconfig.py module that now
-        # exists as part of the sonic-config-engine package.
-        title = []
+        # XXX: move the porttab
+        # parsing stuff to a separate module, or reuse
+        # if something already exists
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Please change this comment back, as the other comment is newer and more descriptive as to what needs to be done.

File: sonic_sfp/sfputilbase.py [hunk 5]
```diff
@@ -268,15 +310,14 @@ def read_porttab_mappings(self, porttabfile):
...
         for line in f:
             line.strip()
+            title = []
             if re.search("^#", line) is not None:
                 # The current format is: # name lanes alias index speed
-                # Where the ordering of the columns can vary
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Why remove this line?

File: sonic_sfp/sfputilbase.py [hunk 6]
```diff
@@ -582,6 +623,179 @@ def get_eeprom_dict(self, port_num):
...
+
+        sfp_vendor_sn_raw = self._read_eeprom_specific_bytes(sysfsfile_eeprom, (offset + XCVR_VENDOR_SN_OFFSET), XCVR_VENDOR_SN_WIDTH)
+        if sfp_vendor_sn_raw is not None:
+            sfp_vendor_sn_data = sfpi_obj.parse_vendor_sn(sfp_vendor_sn_raw, 0)
+        else:
+            return None
+
+        try:
+            sysfsfile_eeprom.close()
+        except:
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: `except IOError:` ?

File: sonic_sfp/sfputilbase.py [hunk 7]
```diff
@@ -582,6 +623,179 @@ def get_eeprom_dict(self, port_num):
...
+            print("Error: file close failed")
+            return None
+
+        transceiver_info_dict['type'] = sfp_type_data['data']['type']['value']
+        transceiver_info_dict['manufacturename'] = sfp_vendor_name_data['data']['Vendor Name']['value']
+        transceiver_info_dict['modelname'] = sfp_vendor_pn_data['data']['Vendor PN']['value']
+        transceiver_info_dict['hardwarerev'] = sfp_vendor_rev_data['data']['Vendor Rev']['value']
+        transceiver_info_dict['serialnum'] = sfp_vendor_sn_data['data']['Vendor SN']['value']
+
+        #print("transceiver info succcessfully fetched")
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Remove commented debug print statement.

File: sonic_sfp/sfputilbase.py [hunk 8]
```diff
@@ -582,6 +623,179 @@ def get_eeprom_dict(self, port_num):
...
+        return transceiver_info_dict
+        #return sfp_type_data, sfp_vendor_name_data, sfp_vendor_pn_data, sfp_vendor_rev_data, sfp_vendor_sn_data
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Remove?

File: sonic_sfp/sfputilbase.py [hunk 9]
```diff
@@ -582,6 +623,179 @@ def get_eeprom_dict(self, port_num):
...
+            transceiver_dom_info_dict['rx3power'] = 'N/A'
+            transceiver_dom_info_dict['rx4power'] = 'N/A'
+            transceiver_dom_info_dict['tx1bias'] = dom_channel_monitor_data['data']['TXBias']['value']
+            transceiver_dom_info_dict['tx2bias'] = 'N/A'
+            transceiver_dom_info_dict['tx3bias'] = 'N/A'
+            transceiver_dom_info_dict['tx4bias'] = 'N/A'
+
+        try:
+            sysfsfile_eeprom.close()
+        except:
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: `except IOError:` ?

File: sonic_sfp/sfputilbase.py [hunk 10]
```diff
@@ -582,6 +623,179 @@ def get_eeprom_dict(self, port_num):
...
+            return None
+
+        transceiver_dom_info_dict['temperature'] = dom_temperature_data['data']['Temperature']['value']
+        transceiver_dom_info_dict['voltage'] = dom_voltage_data['data']['Vcc']['value']
+
+        return transceiver_dom_info_dict
+        #return dom_temperature_data, dom_voltage_data, dom_channel_monitor_data
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Remove?

File: sonic_sfp/sfputilbase.py [hunk 11]
```diff
@@ -614,3 +828,13 @@ def reset(self, port_num):
         :returns: Boolean, True if reset successful, False if not
         """
         return
+
+    @abc.abstractmethod
+    def get_transceiver_change_event(self, timeout=0):
+        """
+        :param timeout
+		:returns: Boolean, True if call successful, False if not;
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Replace tabs with spaces.

File: sonic_sfp/sfputilbase.py [hunk 12]
```diff
@@ -614,3 +828,13 @@ def reset(self, port_num):
...
+        dict for pysical interface number and the SFP status,
+		status='1' represent plug in, '0' represent plug out like {'0': '1', '31':'0'}
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Replace tabs with spaces.

File: sonic_sfp/sff8472.py [hunk 13]
```diff
@@ -1025,6 +1076,40 @@ def calc_rx_power(self, eeprom_data, offset, size):
...
+    dom_channel_monitor_params = {
+        'TXBias':
+            {'offset':0,
+             'size':2,
+             'type':'func',
+             'decode':{'func':calc_bias}},
+        'TXPower':
+            {'offset':2,
+             'size':2,
+             'type':'func',
```
Comment (hui-ma) [inline sonic_sfp/sff8472.py]: Could we expose TXpower from the sfputilbase and add them into Database?

File: sonic_sfp/sfputilbase.py [hunk 14]
```diff
@@ -714,6 +720,20 @@ def get_transceiver_dom_info_dict(self, port_num):
             if sfpd_obj is None:
                 return None
 
+            sfpi_obj = sff8436InterfaceId()
+            if sfpi_obj is None:
+                return None
+
+            # QSFP capability byte parse, through this byte can know whether it support tx_power or not.
+            # TODO: in the future when decided to migarate to support SFF-8636 instead of SFF-8436,
```
Comment (jleveque) [inline sonic_sfp/sfputilbase.py]: Typo: "migarate" => "migrate"

**Pull request merged**